File: src/rbry/runlib/interface/in.ts
import { getClass } from "../helpers/getClass";

export default function _in(e: any, c: any) {

    const kc = getClass(c);
    if( kc === undefined ) {
        console.warn(c);
        throw new Error("?")
    }

    let fct = kc.prototype["__contains__"];
    if( fct !== undefined )
        return fct.call(c, e);

    fct = kc.prototype[Symbol.iterator] //kc.prototype["__iter__"];
    if( fct !== undefined ) {
        for( let elem of fct.call(c) )
            // @ts-ignore
            if( $RB.op(elem, "eq", e) )
                return true;
        return false;
    }

    throw new Error("Not implemented");
    //TODO: getitem...

}

File: src/tests/py/kernel/op/in.py
class X:
    def __contains__(self, o: object):
        return o is 1

x = X()
assert 1 in x
assert 2 not in x

class Y:
    def __iter__(self):
        yield 1

y = Y()
assert 1 in y # pyright: ignore[reportOperatorIssue]
assert 2 not in y # pyright: ignore[reportOperatorIssue]